packages/react-aria-components/src/Input.tsx

```diff
@@ -61,6 +61,10 @@ function Input(props: InputProps, ref: ForwardedRef<HTMLInputElement>) {
       {...mergeProps(props, focusProps, hoverProps)}
       {...renderProps}
       ref={ref}
+      data-focused={isFocused || undefined}
+      data-disabled={props.disabled || undefined}
+      data-readonly={props.readOnly || undefined}
+      data-invalid={!!props['aria-invalid'] || undefined}
```

Comment:
readonly and invalid don't seem to be documented above or included in the `useRenderProps` values.